### PR TITLE
Report obsolete diagnostics on `System.Index` and `System.Range`

### PR DESCRIPTION
**Pull request by DoctorKrolic**
Fixes: https://github.com/dotnet/roslyn/issues/57745

### CODE REVIEW

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 1]
```diff
@@ -2597,6 +2602,7 @@ private BoundExpression BindRangeExpression(RangeExpressionSyntax node, BindingD
             CheckFeatureAvailability(node, MessageID.IDS_FeatureRangeOperator, diagnostics);
 
             TypeSymbol rangeType = GetWellKnownType(WellKnownType.System_Range, diagnostics, node);
+            ReportDiagnosticsIfObsolete(diagnostics, rangeType, node, hasBaseReceiver: false);
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2605]: I looked around for other references to `WellKnownType.System_Range/System_Index` to see if more scenarios have a similar issue.
The ones related to implicit index/range indexers seem fine.
But a few more places need closer look (likely are failing to report obsolete diagnostic):
- slice patterns (see `BindSlicePattern`),
- inline array index (see `tryImplicitConversionToInlineArrayIndex` and `bindInlineArrayElementAccess`). It looks like the scenario may require an implicit user-defined conversion from the operand type to `System.Index` or `System.Range`
- `BindRangeExpressionOperand` uses `WellKnownType.System_Index`
Could you add a test for these scenarios and either fix them or file a tracking issue?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 2]
```diff
@@ -2588,6 +2589,10 @@ private BoundExpression BindFromEndIndexExpression(PrefixUnaryExpressionSyntax n
 
             BoundExpression boundConversion = CreateConversion(boundOperand, conversion, intType, diagnostics);
             MethodSymbol symbolOpt = GetWellKnownTypeMember(WellKnownMember.System_Index__ctor, diagnostics, syntax: node) as MethodSymbol;
+            if (symbolOpt is not null)
+            {
+                ReportDiagnosticsIfObsolete(diagnostics, symbolOpt, node, hasBaseReceiver: false);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:2594]: This should probably also check `UnmanagedCallersOnly`

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1). Looks good but may be some more interesting scenarios to test

**Comment by DoctorKrolic** [comment]
@jcouv @333fred A thought came to my mind: what if we move "report if obsolete" logic down to `GetWellKnownType`/`GetWellKnownTypeMember`? They already receive a syntax node to report a use site diagnostics on, well, they can use this node to also eportobsolete diagnostics as well. This way such logic will implicitly apply to new well-known types/members as they are added to the compiler and we want such behaviour, right?

**Comment by jcouv** [comment]
> what if we move "report if obsolete" logic down to GetWellKnownType/GetWellKnownTypeMember?

Discussed further with @333fred and @AlekseyTs following this question. 
It would indeed be beneficial to simplify the code and be more systematic, but the churn, review and possible complications don't seem worthwhile.
Going in the other direction, we think it would be better to close this category of issues and make it clear that we don't generally not bother reporting Obsolete/UnmanagedCallersOnly diagnostics for well-known types/members, unless there is a strong reason to.
I'll do a search to close such issues. I also took a stab at some [documentation](https://github.com/dotnet/roslyn/pull/75522) to clarify this for future work.

As far as the present PR, I would recommend closing it. Sorry for the trouble

**Pull request closed without merging**